x-pack/plugins/cross_cluster_replication/__jest__/client_integration/helpers/auto_follow_pattern_add.helpers.js

```diff
@@ -0,0 +1,34 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import { registerTestBed } from '../../../../../test_utils';
+import { AutoFollowPatternAdd } from '../../../public/app/sections/auto_follow_pattern_add';
+import { ccrStore } from '../../../public/app/store';
+import routing from '../../../public/app/services/routing';
+
+const testBedOptions = {
+  memoryRouter: {
+    onRouter: (router) => routing.reactRouter = router
+  }
+};
+
+const initTestBed = registerTestBed(AutoFollowPatternAdd, { options: testBedOptions, store: ccrStore });
+
+export const setup = (props) => {
```

Comment:
Thanks to Typescript... all your concerns will disappear 😊 

There is a tendency in Kibana in splitting too quick in multiple files the code and IMO it makes it very hard to follow, and discovery is almost inexistent.

The `testbed.ts` file is 197 lines long (30% of it comments). In that file, I have everything I need as a developer to understand what it does without the need to jump through multiple files. If it ever grows into this giant file you mention we could split it. But I think there is still some room.

You mention that for you the current testBed is opake...  Not with Typescript:

<img width="652" alt="Screen Shot 2019-04-24 at 07 29 05" src="https://user-images.githubusercontent.com/2854616/56634577-c33e9f00-6662-11e9-86c1-26d62a720a07.png">

To me, having loads of files in folders I don't know about is much more opake. And as a new developer joining the team, I find it more explicit and helpful a clear interface like above as to have to parse folders in the repo to see what's available to use in my tests.

Last but not least 😊 You solution is much more verbose, by giving a fixed structure we have all our test files similar (`form`, `table`, `actions`) and we can almost copy paste

```js
let find;
let exists;
let component;
let table;
let actions; // one variable for all the actions in the tests
let tableCellsValues;
let rows;
```

and be ready to test!

With that said, if you really really want those file exported for when you will add some tests, we can export those function in separate files (but still leave them as part of the testbed as I really like the interface as a Developer experience) but we'd lose readability in your tests.

Reply:
Thanks for offering to break out the functions into separate files. I think I'd be happy with simply making the helper files explicit about what they're exposing to the test:

```js
export const setup = (props) => {
  const {
    find,
    exists,
    component,
    form: {
      getErrorsMessages,
      setComboBoxValue,
      setInputValue,
    },
  } = initTestBed(props);

  // User actions
  const clickSaveForm = () => {
    testBed.find('submitButton').simulate('click');
  };
  
  return {
    component,
    find,
    exist,
    getErrorsMessages,
    setComboBoxValue,
    setInputValue,
    clickSaveForm,
  };
};
```

This way, when I read the code I can get idea of what types of interactions we're going to perform on the page just from scanning this helper. I'd also prefer to flatten the returned object because I think they add noise, but that's my personal preference.